File: lib/types/intake.ts
import type { BodyRegion } from "./body";

/**
 * Subjective, user-reported pain data. This is the raw input side of the
 * intake — it represents what the user told us, NOT a clinical assessment.
 * The Extractor agent merges this with evidence from uploaded files to
 * produce the normalized ExtractedClinicalPayload.
 */

export type PainQuality =
  | "sharp"
  | "dull"
  | "aching"
  | "throbbing"
  | "burning"
  | "stabbing"
  | "tingling"
  | "pins_and_needles"
  | "stiffness";

export type PainOnset =
  | "sudden_traumatic"
  | "sudden_atraumatic"
  | "gradual"
  | "chronic_recurring"
  | "unknown";

export type Aggravator =
  | "flexion"
  | "extension"
  | "rotation"
  | "lateral_flexion"
  | "weight_bearing"
  | "sitting_prolonged"
  | "standing_prolonged"
  | "walking"
  | "running"
  | "lifting"
  | "overhead_reach"
  | "gripping"
  | "cold_weather"
  | "morning_stiffness"
  | "end_of_day"
  | "after_exercise"
  | "at_rest"
  | "at_night";

export type Reliever =
  | "rest"
  | "heat"
  | "ice"
  | "movement"
  | "stretching"
  | "nsaids"
  | "position_change"
  | "nothing_helps";

export interface SubjectivePainInput {
  /** Primary site the user identifies as the worst pain. */
  primaryLocation: BodyRegion | null;
  /** Optional additional sites, in user-reported order. */
  secondaryLocations: BodyRegion[];
  /** 0–10 visual analog scale. Null until user interacts. */
  intensity: number | null;
  qualities: PainQuality[];
  onset: PainOnset | null;
  /** Null if unknown; 0 allowed for "started today". */
  durationDays: number | null;
  aggravators: Aggravator[];
  relievers: Reliever[];
  /**
   * Free-text subjective description. Gets streamed into the Triage agent
   * verbatim for red-flag keyword scanning BEFORE the Extractor runs.
   */
  freeText: string;
}

export const EMPTY_SUBJECTIVE_INPUT: SubjectivePainInput = {
  primaryLocation: null,
  secondaryLocations: [],
  intensity: null,
  qualities: [],
  onset: null,
  durationDays: null,
  aggravators: [],
  relievers: [],
  freeText: ""
};

/**
 * Metadata for a file the user uploaded. The file bytes themselves live in
 * object storage (backend-side) — we keep only the pointer and provenance
 * here so the client state is never holding PHI blobs past upload.
 */
export type UploadedFileCategory =
  | "imaging_report"   // radiology narrative, PDF
  | "imaging_dicom"    // raw imaging (MRI/CT/X-ray)
  | "imaging_photo"    // user-captured photo of a printed scan
  | "clinical_note"    // referral letter, consult note
  | "lab_report"
  | "other";

export type UploadStatus =
  | "queued"
  | "uploading"
  | "uploaded"
  | "processing"   // handed off to Extractor/VLM
  | "failed";

export interface UploadedFileMeta {
  /** Client-side UUID — stable across the upload lifecycle. */
  id: string;
  filename: string;
  mimeType: string;
  sizeBytes: number;
  category: UploadedFileCategory;
  uploadedAt: string; // ISO 8601
  status: UploadStatus;
  /** Server-issued opaque key, populated once the upload completes. */
  storageKey?: string;
  /** User-supplied note (e.g., "MRI from April 2025"). */
  note?: string;
  /** Set when status === "failed". */
  error?: string;
}

/**
 * Top-level intake record. Carries the full user-reported picture plus the
 * outputs of each agent as they run. The `status` field is the HITL
 * state machine — no exercise plan is surfaced until status reaches
 * `plan_ready` AND a clinician has cleared it (handled server-side).
 */
export type IntakeStatus =
  | "draft"                     // user still filling the form
  | "submitted"                 // sent to orchestrator
  | "triaging"                  // Agent 2 running
  | "halted_red_flag"           // Agent 2 hit a critical keyword — no plan
  | "extracting"                // Agent 1 running on files
  | "retrieving"                // Agent 3 querying RAG
  | "pending_clinical_review"   // plan drafted, awaiting human sign-off
  | "plan_ready"                // released to user
  | "failed";

export interface IntakeSubmission {
  id: string;
  createdAt: string;            // ISO 8601
  status: IntakeStatus;
  subjective: SubjectivePainInput;
  files: UploadedFileMeta[];
  /** Filled by the agent pipeline; all optional here. */
  extracted?: import("./agents").ExtractedClinicalPayload;
  triage?: import("./agents").TriageResult;
  plan?: import("./agents").RehabPlan;
}

File: backend/app/schemas/intake.py
from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING

from pydantic import BaseModel, Field

from app.schemas.body import BodyRegion

if TYPE_CHECKING:
    from app.schemas.agents import (
        ExtractedClinicalPayload,
        RehabPlan,
        TriageResult,
    )


class PainQuality(StrEnum):
    sharp = "sharp"
    dull = "dull"
    aching = "aching"
    throbbing = "throbbing"
    burning = "burning"
    stabbing = "stabbing"
    tingling = "tingling"
    pins_and_needles = "pins_and_needles"
    stiffness = "stiffness"


class PainOnset(StrEnum):
    sudden_traumatic = "sudden_traumatic"
    sudden_atraumatic = "sudden_atraumatic"
    gradual = "gradual"
    chronic_recurring = "chronic_recurring"
    unknown = "unknown"


class Aggravator(StrEnum):
    flexion = "flexion"
    extension = "extension"
    rotation = "rotation"
    lateral_flexion = "lateral_flexion"
    weight_bearing = "weight_bearing"
    sitting_prolonged = "sitting_prolonged"
    standing_prolonged = "standing_prolonged"
    walking = "walking"
    running = "running"
    lifting = "lifting"
    overhead_reach = "overhead_reach"
    gripping = "gripping"
    cold_weather = "cold_weather"
    morning_stiffness = "morning_stiffness"
    end_of_day = "end_of_day"
    after_exercise = "after_exercise"
    at_rest = "at_rest"
    at_night = "at_night"


class Reliever(StrEnum):
    rest = "rest"
    heat = "heat"
    ice = "ice"
    movement = "movement"
    stretching = "stretching"
    nsaids = "nsaids"
    position_change = "position_change"
    nothing_helps = "nothing_helps"


class UploadedFileCategory(StrEnum):
    imaging_report = "imaging_report"
    imaging_dicom = "imaging_dicom"
    imaging_photo = "imaging_photo"
    clinical_note = "clinical_note"
    lab_report = "lab_report"
    other = "other"


class UploadStatus(StrEnum):
    queued = "queued"
    uploading = "uploading"
    uploaded = "uploaded"
    processing = "processing"
    failed = "failed"


class SubjectivePainInput(BaseModel):
    primary_location: BodyRegion | None = Field(default=None, alias="primaryLocation")
    secondary_locations: list[BodyRegion] = Field(default_factory=list, alias="secondaryLocations")
    intensity: int | None = None
    qualities: list[PainQuality] = Field(default_factory=list)
    onset: PainOnset | None = None
    duration_days: int | None = Field(default=None, alias="durationDays")
    aggravators: list[Aggravator] = Field(default_factory=list)
    relievers: list[Reliever] = Field(default_factory=list)
    free_text: str = Field(default="", alias="freeText")

    model_config = {"populate_by_name": True}


class UploadedFileMeta(BaseModel):
    id: str
    filename: str
    mime_type: str = Field(alias="mimeType")
    size_bytes: int = Field(alias="sizeBytes")
    category: UploadedFileCategory
    uploaded_at: datetime = Field(alias="uploadedAt")
    status: UploadStatus
    storage_key: str | None = Field(default=None, alias="storageKey")
    note: str | None = None
    error: str | None = None

    model_config = {"populate_by_name": True}


class IntakeStatus(StrEnum):
    draft = "draft"
    submitted = "submitted"
    triaging = "triaging"
    halted_red_flag = "halted_red_flag"
    extracting = "extracting"
    retrieving = "retrieving"
    pending_clinical_review = "pending_clinical_review"
    plan_ready = "plan_ready"
    failed = "failed"


class IntakeSubmission(BaseModel):
    """Top-level submission record. Stored server-side; `plan` stays None
    until retrieval AND clinician-review complete."""

    id: str
    created_at: datetime = Field(alias="createdAt")
    status: IntakeStatus
    subjective: SubjectivePainInput
    files: list[UploadedFileMeta] = Field(default_factory=list)

    # Populated by the agent pipeline
    extracted: "ExtractedClinicalPayload | None" = None
    triage: "TriageResult | None" = None
    plan: "RehabPlan | None" = None

    model_config = {"populate_by_name": True}


class SubmitIntakeRequest(BaseModel):
    subjective: SubjectivePainInput
    files: list[UploadedFileMeta] = Field(default_factory=list)
    consent_version: str = Field(alias="consentVersion")

    model_config = {"populate_by_name": True}


class SubmitIntakeResponse(BaseModel):
    submission_id: str = Field(alias="submissionId")
    triage: "TriageResult"
    halted: bool
    poll_after_seconds: int | None = Field(alias="pollAfterSeconds")

    model_config = {"populate_by_name": True}


# Rebuild to resolve forward refs once the agents module has loaded.
def _rebuild() -> None:
    from app.schemas.agents import (  # noqa: F401
        ExtractedClinicalPayload,
        RehabPlan,
        TriageResult,
    )

    IntakeSubmission.model_rebuild()
    SubmitIntakeResponse.model_rebuild()


_rebuild()

